starport-blog/02-cli-list-posts.md

```diff
@@ -0,0 +1,141 @@
+
+## x/blog/client/cli/query.go
+
+```go
+import (
+  // Existing imports...
+  "github.com/cosmos/cosmos-sdk/client/context"
+)
+```
+
+Function GetQueryCmd is used for creating a list of `query` subcommands, it should already be defined. Edit the function to add `GetCmdListPosts` as a subcommand:
+
+```go
+  blogQueryCmd.AddCommand(
+    flags.GetCommands(
+      GetCmdListPosts(queryRoute, cdc),
+    )...,
+  )
+```
+
+Now let’s define `GetCmdListPosts`:
+
+```go
+func GetCmdListPosts(queryRoute string, cdc *codec.Codec) *cobra.Command {
+  return &cobra.Command{
+    Use:   "list-posts",
+    Short: "list all posts",
+    RunE: func(cmd *cobra.Command, args []string) error {
+      cliCtx := context.NewCLIContext().WithCodec(cdc)
+      res, _, err := cliCtx.QueryWithData(fmt.Sprintf("custom/%s/"+types.QueryListPosts, queryRoute), nil)
+      if err != nil {
+        fmt.Printf("could not list posts\n%s\n", err.Error())
+        return nil
+      }
+      var out []types.Post
+      cdc.MustUnmarshalJSON(res, &out)
+      return cliCtx.PrintOutput(out)
+    },
+  }
+}
+```
+
+`GetCmdListPosts` runs an ABCI query to fetch the data, unmarshals it back form binary to JSON and returns it to the console. ABCI is an interface between your app and Tendermint (a program responsible for replicating the state across machines). ABCI queries look like paths on a hierarchical filesystem. In our case, the query is `custom/blog/list-posts`. Before we continue, we need to define `QueryListPosts`.
+
+## x/blog/types/querier.go
+
+```go
+const (
+  QueryListPosts = "list-posts"
+)
+```
+
+## x/blog/client/cli/keeper/querier.go
+
+```go
+import (
+  // Existing imports ...
+  "github.com/example/blog/x/blog/types"
+  "github.com/cosmos/cosmos-sdk/codec"
+)
+```
+
+`NewQuerier` acts as a dispatcher for query functions, it should already be defined. Modify the switch statement to include `listPosts`:
+
+```go
+    switch path[0] {
+    case types.QueryListPosts:
+      return listPosts(ctx, k)
+    default:
+```
+
+Now let’s define `listPosts`:
+
+```go
+func listPosts(ctx sdk.Context, k Keeper) ([]byte, error) {
+  var postList []types.Post
+  store := ctx.KVStore(k.storeKey)
+  iterator := sdk.KVStorePrefixIterator(store, []byte(types.PostPrefix))
+  for ; iterator.Valid(); iterator.Next() {
+    var post types.Post
+    k.cdc.MustUnmarshalBinaryLengthPrefixed(store.Get(iterator.Key()), &post)
+    postList = append(postList, post)
+  }
+  res := codec.MustMarshalJSONIndent(k.cdc, postList)
+  return res, nil
+}
+```
+
+This function uses an prefix iterator to loop through all the keys with a given prefix (in our case `PostPrefix` is `"post-"`). We’re getting values by key with `store.Get` and appending them to postList. Finally, we unmarshal bytes back to JSON and return the result to the console.
+
+Now let’s see how it works. Run the following command to recompile your app, clear the data and relaunch the chain:
```

Comment:
checking back to my console i have this error:
```
2020/07/24 16:19:12 Error in building the application. Please, check ./Makefile
```
Which is good in that the console has stopped so i can re-run `starport serve` but it seems that shouldn't be the case. Should the original `starport serve` have stopped already or should there be an instruction to stop it? Or was this supposed to be part of a hot reload?

Reply:
You probably saved a file, hot-reload tried rebuilding and failed and it ended up with an error. We need to rewrite `starport serve` so that a build error doesn't exit Starport, but rather makes it wait for further changes (sort of like on the frontend).